### ISSUE TITLE
RuntimeError when removing agents from model with new schedulers

### ISSUE THREAD

**Issue by quaquel** [projectmesa/mesa#1959]
**Describe the bug**
WolfSheep and similar models where agents are removed during `agent.step` throw a Runtime Error:

```python
RuntimeError: dictionary changed size during iteration
```

The cause is quite straightforward: in `AgentSet.do`, one iterates over the agents. If `do` triggers the removal of agents that are in the AgentSet over which you are iterating, you get the error. 

**Expected behavior**
The Runtime error should not occur

**Proposed fix**
Change line 254 in agent.py to `res = [getattr(agent, method_name)(*args, **kwargs) for agent in list(self._agents)]`. You no longer get the error by turning the keys into a list and iterating over this list. This is also identical to how the error was prevented in the [2.1.5 version of the schedulers](https://github.com/projectmesa/mesa/blob/53e8761b47109b18ff77fae78a9024376b14d0e2/mesa/time.py#L106-L112). 

In addition, it would be good to also add an explicit unit test for this error so that it won't happen in the future. Something like below should be all that is needed

```python

MockAgent(agent):
    def step(self):
        self.model.remove()

agents = [MockAgent(i, mockmodel) for i in range(10)]
agentset = AgentSet(agents, mockmodel)
agentset.do("step")

```

**To Reproduce**
Run WolfSheep with the latest stable version of MESA

**Comment by rht** [comment]
An attempt to fix was made at #1958.

**Comment by quaquel** [comment]
I have done some additional diagnostics on this.

Specifically, with respect to the WolfSheep model, there are two ways in which the scheduler can be changed while iterating over it. First, the error happens because a sheep or wolf is born and added to the model. Second,  the error happens when a wolf or sheep dies and thus is removed from the model.

WeakkeyDict runs with a form of protection. See [StackOverflow](https://stackoverflow.com/questions/12428026/safely-iterating-over-weakkeydictionary-and-weakvaluedictionary#:~:text=It%20is%20actually%20safe%20to,the%20docs%20never%20got%20updated.) for more details. If I change, in the example model, the code from `self.model.schedule.remove` to `self.remove`, this prevents errors because a wolf or sheep has died. However, this proper fix is not backward compatible and can only be relied upon partially.

What remains, thus, is the addition of new agents while iterating and casting the keys to list or copying them before iterating solves this problem but can trigger other errors as discussed in #1958.

In short, the error is triggered when explicitly calling `AgentSet.add` and `AgentSet.remove` while iterating over the AgentSet.

**Comment by quaquel** [comment]
closed by #1960
